libethereum/BlockChain.cpp

```diff
@@ -226,88 +222,107 @@ bool BlockChain::open(fs::path const& _path, WithExisting _we)
+                       "details. Exception: "
                     << ex.what();
                 BOOST_THROW_EXCEPTION(DatabaseCorruption());
             }
             else if (dbStatus == db::DatabaseStatus::IOError)
             {
-                cerror << "Database already open. You appear to have another instance of Aleth running.";
+                LOG(m_loggerError) << "Database already open. You appear to have another instance "
+                                      "of Aleth running.";
                 BOOST_THROW_EXCEPTION(DatabaseAlreadyOpen());
             }
         }
-        
-        cerror << "Unknown error occurred. Exception details: " << ex.what();
+
+        LOG(m_loggerError) << "Unknown error occurred when opening database. Exception details: "
+                           << ex.what();
         throw;
     }
 
     if (_we != WithExisting::Verify && !rebuildNeeded && !details(m_genesisHash))
     {
-        BlockHeader gb(m_params.genesisBlock());
+        bytes const genesisBlockBytes = m_params.genesisBlock();
+        BlockHeader const genesisHeader{genesisBlockBytes};
         // Insert details of genesis block.
-        m_details[m_genesisHash] = BlockDetails(0, gb.difficulty(), h256(), {});
-        auto r = m_details[m_genesisHash].rlp();
-        m_extrasDB->insert(toSlice(m_genesisHash, ExtraDetails), (db::Slice)dev::ref(r));
-        assert(isKnown(gb.hash()));
+        m_details[m_genesisHash] = BlockDetails{0 /* number */, genesisHeader.difficulty(),
+            h256{} /* parent */, {} /* children */, genesisBlockBytes.size()};
+        auto const genesisDetailsRlp = m_details[m_genesisHash].rlp();
+        m_extrasDB->insert(
+            toSlice(m_genesisHash, ExtraDetails), (db::Slice)dev::ref(genesisDetailsRlp));
+        assert(isKnown(genesisHeader.hash()));
     }
 
     // TODO: Implement ability to rebuild details map from DB.
     auto const l = m_extrasDB->lookup(db::Slice("best"));
     m_lastBlockHash = l.empty() ? m_genesisHash : h256(l, h256::FromBinary);
+    // We need to retrieve the block number from the blocks database rather than from the extras
+    // database because the extras database format may have changed
+    m_lastBlockNumber = info(m_lastBlockHash).number();
 
-    m_lastBlockNumber = number(m_lastBlockHash);
-
-    ctrace << "Opened blockchain DB. Latest: " << currentHash()
-           << (!rebuildNeeded ? "(rebuild not needed)" : "*** REBUILD NEEDED ***");
+    LOG(m_loggerInfo) << "Opened blockchain database. Latest block hash: " << currentHash()
+                      << (!rebuildNeeded ? "(rebuild not needed)" : "*** REBUILD NEEDED ***");
     return rebuildNeeded;
 }
 
 void BlockChain::open(fs::path const& _path, WithExisting _we, ProgressCallback const& _pc)
 {
     if (open(_path, _we) || _we == WithExisting::Verify)
-        rebuild(_path, _pc);
+        rebuild(fs::path{}, _pc);
```

Comment:
@gumb0: Wanted a way to avoid unnecessarily reconstructing the DatabasePaths object...I don’t think we need to recreate it in rebuild if we already have built m_dbPaths